pkg/adapter/awscognitouserpoolsource/adapter_test.go

```diff
@@ -0,0 +1,162 @@
+)
+
+type mockedCognitoUserPoolClient struct {
+	cognitoidentityprovideriface.CognitoIdentityProviderAPI
+	listUsersOutput cognitoidentityprovider.ListUsersOutput
+}
+
+func (m mockedCognitoUserPoolClient) ListUsers(in *cognitoidentityprovider.ListUsersInput) (*cognitoidentityprovider.ListUsersOutput, error) {
+	return &m.listUsersOutput, nil
+}
+
+func TestListUsers(t *testing.T) {
+	user := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: aws.Time(time.Now()),
+	}
+
+	a := &adapter{
+		userPoolID: "userpool/fooPool",
+		logger:     loggingtesting.TestLogger(t),
+		ceClient:   adaptertest.NewTestClient(),
+		cgnIdentityClient: mockedCognitoUserPoolClient{
+			listUsersOutput: cognitoidentityprovider.ListUsersOutput{
+				Users: []*cognitoidentityprovider.UserType{
+					&user,
+				},
+			},
+		},
+	}
+
+	users, err := a.listUsers()
+	assert.NoError(t, err)
+	assert.Equal(t, []*cognitoidentityprovider.UserType{&user}, users)
+}
+
+func TestFilterByTimestamp(t *testing.T) {
+	now := time.Unix(0, 0)
+
+	// existing user
+	user1 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: &now,
+	}
+
+	// we must get 0 new users
+	users, timestamp := filterByTimestamp([]*cognitoidentityprovider.UserType{
+		&user1,
+	}, now)
+	assert.Len(t, users, 0)
+
+	// new user
+	user2 := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: aws.Time(time.Unix(0, 1)),
+	}
+
+	// we must get user2 only
+	users, timestamp = filterByTimestamp([]*cognitoidentityprovider.UserType{
+		&user1,
+		&user2,
+	}, timestamp)
+	assert.Equal(t, []*cognitoidentityprovider.UserType{&user2}, users)
+
+	// repeat filtering, we must ignore user2
+	users, _ = filterByTimestamp([]*cognitoidentityprovider.UserType{
+		&user1,
+		&user2,
+	}, timestamp)
+	assert.Len(t, users, 0)
+}
+
+func TestSendCognitoEvent(t *testing.T) {
+	ceClient := adaptertest.NewTestClient()
+
+	user := cognitoidentityprovider.UserType{
+		UserLastModifiedDate: aws.Time(time.Now().UTC()),
+	}
+
+	a := &adapter{
+		userPoolID:        "userpool/fooPool",
+		logger:            loggingtesting.TestLogger(t),
+		ceClient:          ceClient,
+		cgnIdentityClient: mockedCognitoUserPoolClient{},
+	}
+
+	err := a.sendCognitoEvent(&user)
+	assert.NoError(t, err)
+
+	events := ceClient.Sent()
+	assert.Len(t, events, 1)
+
+	var gotUser cognitoidentityprovider.UserType
+	err = events[0].DataAs(&gotUser)
+	assert.NoError(t, err)
+	// make sure that we get what we sent
+	assert.Equal(t, user, gotUser)
+}
+
+func TestStart(t *testing.T) {
+	testTimeout := 2 * time.Second
+
+	a := &adapter{
+		userPoolID:        "userpool/fooPool",
+		logger:            loggingtesting.TestLogger(t),
+		ceClient:          adaptertest.NewTestClient(),
+		cgnIdentityClient: mockedCognitoUserPoolClient{},
+	}
+
+	// create a context that will be done in 2s
+	testCtx, testCancel := context.WithTimeout(context.Background(), testTimeout)
+	defer testCancel() // not really necessary here, but let's follow good practices
```

Comment:
That comment was for the review 🙂   It can stay but it's a little bit unnecessary.